### PR TITLE
Optimise docker image build: copy python requirements file only before install them

### PR DESCRIPTION
**Pull request by zhangyuan**
# Description

When build the docker image, the whole `flow` directory is added to docker image before installing the dependencies (python and debian packages, etc). However any minor changes results in re-installing the dependencies. To avoid the unneccessary time-consuming installation, only copy the `requirements.txt` that's required by `pip install` first, and then install the depedencies.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/data/docker/Dockerfile.jinja2 [hunk 1]
```diff
@@ -10,7 +10,7 @@ FROM docker.io/continuumio/miniconda3:latest
 
 WORKDIR /
 
-COPY ./flow /flow
+COPY ./flow/{{env.python_requirements_txt}} /flow/{{env.python_requirements_txt}}
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/data/docker/Dockerfile.jinja2:13]: shall we do a if check here?

Comment (crazygao) [reply src/promptflow/promptflow/_sdk/data/docker/Dockerfile.jinja2:13]: Can this sentence pass if target forlder doesn't exist?

Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/data/docker/Dockerfile.jinja2:13]: need to consider:

1. `env` does not have a `python_requirements_txt`
2. the file `python_requirements_txt` refers does not exist

**Review by zhengfeiwang — COMMENTED** [review]
this change looks good to me, and shall improve docker build perf, but need to consider some cases with checks

File: src/promptflow/promptflow/_sdk/data/docker/Dockerfile.jinja2 (hunk 1 above)
Comment (zhangyuan) [reply src/promptflow/promptflow/_sdk/data/docker/Dockerfile.jinja2:14]: Updated.

How about just checking `env.python_requirements_txt` before `COPY`? It is checked in line 24 as well.

Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/data/docker/Dockerfile.jinja2:14]: please take a look on the CI workflow result: https://github.com/microsoft/promptflow/actions/runs/7034536381?pr=1297, seems there are some jinja syntax error; for example: 

image

please take a look

File: src/promptflow/promptflow/_sdk/data/docker/Dockerfile.jinja2 [hunk 2]
```diff
@@ -10,7 +10,9 @@ FROM docker.io/continuumio/miniconda3:latest
 
 WORKDIR /
 
-COPY ./flow /flow
+{% if env.python_requirements_txt %}
+COPY ./flow/{{env.python_requirements_txt}} /flow/{{env.python_requirements_txt}}
+{% end %}
 
 # create conda environment
 {% if env.conda_file %}
```
Comment (YingChen1996) [inline src/promptflow/promptflow/_sdk/data/docker/Dockerfile.jinja2:18]: it seems also need handle env.conda_file

Comment (zhangyuan) [reply src/promptflow/promptflow/_sdk/data/docker/Dockerfile.jinja2:18]: correct! fixed.

File: src/promptflow/promptflow/_sdk/data/docker/Dockerfile.jinja2 (hunk 1 above)
Comment (zhangyuan) [reply src/promptflow/promptflow/_sdk/data/docker/Dockerfile.jinja2:14]: fixed. thank you @zhengfeiwang

**Review by zhengfeiwang — APPROVED** [review]

**Review by YingChen1996 — APPROVED** [review]

**Review by crazygao — APPROVED** [review]

**Pull request merged**